File: car/src/products/Car.java
package products;

import java.util.Objects;

public class Car {
    private static int wheels = 4;
    private static int instanceCounter = 0;
    private String hersteller;
    private String typ;
    private Double breite;
    private CarColor color;

    public enum CarColor {
        WHITE(0), BLUE(34), RED(31), BLACK(7), GREEN(32);

        private final int seq;

        CarColor(int i) {
            seq = i;
        }
    }

    public Car(String hersteller, String typ, Double breite, CarColor color) {
        this.hersteller = hersteller;
        this.typ = typ;
        this.breite = breite;
        this.color = color;
        instanceCounter++;
    }

    public void paintCar(CarColor newColor) {
        color = newColor;
        System.out.println("\u001B["+color.seq+"mCar is now " + color + "\u001B[0m");
    }

    public void drive() {
        System.out.println("\u001B[" + color.seq + "mbrumm brumm!\u001B[0m");
    }

    public void paintCar() {
        paintCar(CarColor.BLACK);
    }

    public int getWheels() {
        return wheels;
    }

    public void setWheels(int wheels) {
        Car.wheels = wheels;
    }

    public String getHersteller() {
        return hersteller;
    }

    public void setHersteller(String hersteller) {
        this.hersteller = hersteller;
    }

    public String getTyp() {
        return typ;
    }

    public void setTyp(String typ) {
        this.typ = typ;
    }

    public Double getBreite() {
        return breite;
    }

    public void setBreite(Double breite) {
        this.breite = breite;
    }

    public CarColor getColor() {
        return color;
    }

    public static int getInstanceCounter() {
        return instanceCounter;
    }

    @Override
    public String toString() {
        return "\u001B["+color.seq+"mCar{" +
                "hersteller='" + hersteller + '\'' +
                ", typ='" + typ + '\'' +
                ", breite=" + breite +
                ", color='" + color + '\'' +
                ", wheels=" + wheels +
                '}';
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj) return true;
        if (obj == null || getClass() != obj.getClass()) return false;
        Car car = (Car) obj;
        return Objects.equals(getHersteller(), car.getHersteller()) && Objects.equals(getTyp(), car.getTyp()) && Objects.equals(getBreite(), car.getBreite());
    }

    @Override
    public int hashCode() {
        return Objects.hash(getHersteller(), getTyp(), getBreite());
    }
}

File: carKt/src/Car.kt
import java.net.CacheRequest

fun main() {
    val panamera = Car("Porsche", "Panamera", 1.937, CarColor.WHITE)
    val golf = Car("VW", "Golf", 1.795, CarColor.RED)
    val m3 = Car("BMW", "M3", 1.780, CarColor.BLUE)
    val g63 = Car("Mercedes", "G63", 1.885, CarColor.BLACK)
    val gt = Cabrio("Mercedes", "AMG GT Cabriolet", 2.007, CarColor.GREEN)

    panamera.drive()
    golf.drive()
    m3.drive()
    g63.drive()
    gt.drive()
}

enum class CarColor(val seq: Int) {
    WHITE(0), BLUE(34), RED(31), BLACK(7), GREEN(32)
}

open class Car(private var hersteller: String, private var typ: String, private var breite: Double, private var color: CarColor) {

    companion object {
        private var wheels = 4
        var instanceCounter = 0
            private set
    }

    var wheels: Int
        get() = Companion.wheels
        set(wheels) {
            Companion.wheels = wheels
        }

    init {
        instanceCounter++
    }

    fun paintCar(newColor: CarColor = CarColor.BLACK) {
        color = newColor
        println("Car is now \u001B[${color.seq}m$color\u001B[0m")
    }

    open fun drive() {
        println("\u001B[${color.seq}mbrumm brumm\u001B[0m")
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as Car

        if (hersteller != other.hersteller) return false
        if (typ != other.typ) return false
        if (breite != other.breite) return false
        if (color != other.color) return false

        return true
    }

    override fun hashCode(): Int {
        var result = hersteller.hashCode()
        result = 31 * result + typ.hashCode()
        result = 31 * result + breite.hashCode()
        result = 31 * result + color.hashCode()
        return result
    }

    override fun toString(): String {
        return "\u001B[${color.seq}mCar(hersteller='$hersteller', typ='$typ', breite=$breite, color='$color', wheels=$wheels)\u001B[0m"
    }


}
